File: python/commands/project.py
"""
Project-related command implementations for KiCAD interface
"""

import os
import pcbnew  # type: ignore
import logging
from typing import Dict, Any, Optional

logger = logging.getLogger('kicad_interface')

class ProjectCommands:
    """Handles project-related KiCAD operations"""

    def __init__(self, board: Optional[pcbnew.BOARD] = None):
        """Initialize with optional board instance"""
        self.board = board

    def create_project(self, params: Dict[str, Any]) -> Dict[str, Any]:
        """Create a new KiCAD project"""
        try:
            project_name = params.get("projectName", "New_Project")
            path = params.get("path", os.getcwd())
            template = params.get("template")

            # Generate the full project path
            project_path = os.path.join(path, project_name)
            if not project_path.endswith(".kicad_pro"):
                project_path += ".kicad_pro"

            # Create project directory if it doesn't exist
            os.makedirs(os.path.dirname(project_path), exist_ok=True)

            # Create a new board
            board = pcbnew.BOARD()
            
            # Set project properties
            board.GetTitleBlock().SetTitle(project_name)
            
            # Set current date with proper parameter
            from datetime import datetime
            current_date = datetime.now().strftime("%Y-%m-%d")
            board.GetTitleBlock().SetDate(current_date)

            # If template is specified, try to load it
            if template:
                template_path = os.path.expanduser(template)
                if os.path.exists(template_path):
                    template_board = pcbnew.LoadBoard(template_path)
                    # Copy settings from template
                    board.SetDesignSettings(template_board.GetDesignSettings())
                    board.SetLayerStack(template_board.GetLayerStack())

            # Save the board
            board_path = project_path.replace(".kicad_pro", ".kicad_pcb")
            board.SetFileName(board_path)
            pcbnew.SaveBoard(board_path, board)

            # Create project file
            with open(project_path, 'w') as f:
                f.write('{\n')
                f.write('  "board": {\n')
                f.write(f'    "filename": "{os.path.basename(board_path)}"\n')
                f.write('  }\n')
                f.write('}\n')

            self.board = board

            return {
                "success": True,
                "message": f"Created project: {project_name}",
                "project": {
                    "name": project_name,
                    "path": project_path,
                    "boardPath": board_path
                }
            }

        except Exception as e:
            logger.error(f"Error creating project: {str(e)}")
            return {
                "success": False,
                "message": "Failed to create project",
                "errorDetails": str(e)
            }

    def open_project(self, params: Dict[str, Any]) -> Dict[str, Any]:
        """Open an existing KiCAD project"""
        try:
            filename = params.get("filename")
            if not filename:
                return {
                    "success": False,
                    "message": "No filename provided",
                    "errorDetails": "The filename parameter is required"
                }

            # Expand user path and make absolute
            filename = os.path.abspath(os.path.expanduser(filename))

            # If it's a project file, get the board file
            if filename.endswith(".kicad_pro"):
                board_path = filename.replace(".kicad_pro", ".kicad_pcb")
            else:
                board_path = filename

            # Load the board
            board = pcbnew.LoadBoard(board_path)
            self.board = board

            return {
                "success": True,
                "message": f"Opened project: {os.path.basename(board_path)}",
                "project": {
                    "name": os.path.splitext(os.path.basename(board_path))[0],
                    "path": filename,
                    "boardPath": board_path
                }
            }

        except Exception as e:
            logger.error(f"Error opening project: {str(e)}")
            return {
                "success": False,
                "message": "Failed to open project",
                "errorDetails": str(e)
            }

    def save_project(self, params: Dict[str, Any]) -> Dict[str, Any]:
        """Save the current KiCAD project"""
        try:
            if not self.board:
                return {
                    "success": False,
                    "message": "No board is loaded",
                    "errorDetails": "Load or create a board first"
                }

            filename = params.get("filename")
            if filename:
                # Save to new location
                filename = os.path.abspath(os.path.expanduser(filename))
                self.board.SetFileName(filename)

            # Save the board
            pcbnew.SaveBoard(self.board.GetFileName(), self.board)

            return {
                "success": True,
                "message": f"Saved project to: {self.board.GetFileName()}",
                "project": {
                    "name": os.path.splitext(os.path.basename(self.board.GetFileName()))[0],
                    "path": self.board.GetFileName()
                }
            }

        except Exception as e:
            logger.error(f"Error saving project: {str(e)}")
            return {
                "success": False,
                "message": "Failed to save project",
                "errorDetails": str(e)
            }

    def get_project_info(self, params: Dict[str, Any]) -> Dict[str, Any]:
        """Get information about the current project"""
        try:
            if not self.board:
                return {
                    "success": False,
                    "message": "No board is loaded",
                    "errorDetails": "Load or create a board first"
                }

            title_block = self.board.GetTitleBlock()
            filename = self.board.GetFileName()
            
            return {
                "success": True,
                "project": {
                    "name": os.path.splitext(os.path.basename(filename))[0],
                    "path": filename,
                    "title": title_block.GetTitle(),
                    "date": title_block.GetDate(),
                    "revision": title_block.GetRevision(),
                    "company": title_block.GetCompany(),
                    "comment1": title_block.GetComment(0),
                    "comment2": title_block.GetComment(1),
                    "comment3": title_block.GetComment(2),
                    "comment4": title_block.GetComment(3)
                }
            }

        except Exception as e:
            logger.error(f"Error getting project info: {str(e)}")
            return {
                "success": False,
                "message": "Failed to get project information",
                "errorDetails": str(e)
            }

File: src/tools/project.ts
/**
 * Project management tools for KiCAD MCP server
 *
 * These tools handle KiCAD project operations like creating, opening, and saving projects
 */
import { z } from 'zod';
import { logger } from '../logger.js';
import { McpServer } from '@modelcontextprotocol/sdk/server/mcp.js';

// Command function type for KiCAD script calls
type CommandFunction = (command: string, params: Record<string, unknown>) => Promise<any>;

/**
 * Register project management tools with the MCP server
 *
 * @param server MCP server instance
 * @param callKicadScript Function to call KiCAD script commands
 */
export function registerProjectTools(server: McpServer, callKicadScript: CommandFunction) {
    logger.info('Registering project management tools');
    // ------------------------------------------------------
    // Create Project Tool
    // ------------------------------------------------------
    server.tool("create_project", {
        projectName: z.string().describe("Name of the project"),
        path: z.string().describe("Directory path where the project should be created"),
        template: z.string().optional().describe("Optional template to use for the new project")
    }, async ({ projectName, path, template }) => {
        logger.debug(`Creating new project: ${projectName} at ${path}`);
        const result = await callKicadScript("create_project", { projectName, path, template });
        return {
            content: [{
                    type: "text",
                    text: JSON.stringify(result)
                }]
        };
    });
    // ------------------------------------------------------
    // Open Project Tool
    // ------------------------------------------------------
    server.tool("open_project", {
        filename: z.string().describe("Path to the KiCAD project file (.kicad_pro)")
    }, async ({ filename }: { filename: string }) => {
        logger.debug(`Opening project: ${filename}`);
        const result = await callKicadScript("open_project", { filename });
        return {
            content: [{
                    type: "text",
                    text: JSON.stringify(result)
                }]
        };
    });
    // ------------------------------------------------------
    // Save Project Tool
    // ------------------------------------------------------
    server.tool("save_project", {
        filename: z.string().optional().describe("Optional path to save the project to (if different from current)")
    }, async ({ filename }) => {
        logger.debug(`Saving project${filename ? ` to ${filename}` : ''}`);
        const result = await callKicadScript("save_project", { filename });
        return {
            content: [{
                    type: "text",
                    text: JSON.stringify(result)
                }]
        };
    });
    // ------------------------------------------------------
    // Get Project Info Tool
    // ------------------------------------------------------
    server.tool("get_project_info", {}, async () => {
        logger.debug('Getting project information');
        const result = await callKicadScript("get_project_info", {});
        return {
            content: [{
                    type: "text",
                    text: JSON.stringify(result)
                }]
        };
    });
    // ------------------------------------------------------
    // Set Project Properties Tool
    // ------------------------------------------------------
    server.tool("set_project_properties", {
        title: z.string().optional().describe("Project title"),
        company: z.string().optional().describe("Company name"),
        revision: z.string().optional().describe("Project revision"),
        date: z.string().optional().describe("Project date"),
        comment1: z.string().optional().describe("Comment 1"),
        comment2: z.string().optional().describe("Comment 2"),
        comment3: z.string().optional().describe("Comment 3"),
        comment4: z.string().optional().describe("Comment 4")
    }, async (properties: { title?: string, company?: string, revision?: string, date?: string, comment1?: string, comment2?: string, comment3?: string, comment4?: string }) => {
        logger.debug('Setting project properties');
        const result = await callKicadScript("set_project_properties", properties);
        return {
            content: [{
                    type: "text",
                    text: JSON.stringify(result)
                }]
        };
    });
    // ------------------------------------------------------
    // Import Project Tool
    // ------------------------------------------------------
    server.tool("import_project", {
        filename: z.string().describe("Path to the source project file"),
        format: z.enum(["eagle", "altium", "orcad"]).describe("Format of the source project"),
        outputPath: z.string().describe("Directory path where the KiCAD project should be created")
    }, async ({ filename, format, outputPath }: { filename: string, format: string, outputPath: string }) => {
        logger.debug(`Importing ${format} project from ${filename}`);
        const result = await callKicadScript("import_project", {
            filename,
            format,
            outputPath
        });
        return {
            content: [{
                    type: "text",
                    text: JSON.stringify(result)
                }]
        };
    });
    // ------------------------------------------------------
    // Create Backup Tool
    // ------------------------------------------------------
    server.tool("create_backup", {
        backupPath: z.string().optional().describe("Optional directory path for the backup (default: project directory)")
    }, async ({ backupPath }: { backupPath?: string }) => {
        logger.debug(`Creating project backup${backupPath ? ` at ${backupPath}` : ''}`);
        const result = await callKicadScript("create_backup", { backupPath });
        return {
            content: [{
                    type: "text",
                    text: JSON.stringify(result)
                }]
        };
    });
    // ------------------------------------------------------
    // Archive Project Tool
    // ------------------------------------------------------
    server.tool("archive_project", {
        outputPath: z.string().describe("Path where the archive should be created"),
        includeLibraries: z.boolean().optional().describe("Whether to include project libraries"),
        include3dModels: z.boolean().optional().describe("Whether to include 3D models")
    }, async ({ outputPath, includeLibraries, include3dModels }: { outputPath: string, includeLibraries?: boolean, include3dModels?: boolean }) => {
        logger.debug(`Archiving project to ${outputPath}`);
        const result = await callKicadScript("archive_project", {
            outputPath,
            includeLibraries,
            include3dModels
        });
        return {
            content: [{
                    type: "text",
                    text: JSON.stringify(result)
                }]
        };
    });
    logger.info('Project management tools registered');
}
